---
title: "Comprehensive benchmarking of motif instances (all types) - only IntAct and Vidal"
author: "Vitalii Kleshchevnikov"
date: "18/10/2017"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(fig.width = 12, fig.height = 8, warning = FALSE, message = FALSE)

packages = c("MItools", "RColorBrewer", "devtools", "grid","gridExtra", "ggplot2", "VennDiagram", "svglite")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    # specifying mirror is necessary for some Linux systems
    install.packages(packages_to_install, dependencies = T, repos = "http://mirrors.ebi.ac.uk/CRAN/")
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    source("https://bioconductor.org/biocLite.R")
    biocLite(packages_to_install)
    devtools::install_github("vitkl/MItools", dependencies = T)
}
suppressPackageStartupMessages({
    library(MItools)
    library(grid)
    library(gridExtra)
    library(ggplot2)
})

colors = RColorBrewer::brewer.pal(2, "Dark2")

motif_types = c("MOD", "LIG", "DOC") # "MOD", "LIG", "DOC" / "DEG", "CLV", "TRG"
normalise = F # if normalised Sig threshold of 0.3 becomes 1 (but this equalises Sig == 0.3 to true 1 - that is when motif was not found)
#ROC:
measure1 = "prec" # "tpr" or "prec"
measure2 = "rec" # "fpr" or "rec"
both_metrics_vs_cutoff = T
single_metric = c("auc","prbe")[1]
single_metric_name = c("Median AUC", "Prec-rec break-even")[1]

datasets = c("qslimfinder.Full_IntAct3.FALSE",
             "qslimfinder.Vidal3.FALSE",
             "qslimfinder.all_viral_interaction3.FALSE")
descriptions = c("human network (full IntAct) searched \nfor motifs present in viral proteins",
                 "human network (Vidal's data only) searched \nfor motifs present in viral proteins",
                 "human network (human-viral data only) searched \nfor motifs present in viral proteins")
motif_setup_month = c("201802",
                      "201802",
                      "201802")
```

## Overview

### 
1. filter_by_domain_data: NULL or "p.value < 0.5"
2. neg_set: "all_instances", "all_proteins"
3. motif_pval_cutoff: 1, precision == recall, precision > 0.5

```{r, not_filt_Sig}
#####################################################################################
myBenchmarkMotifs = function(neg_set, filter_by_domain_data, motif_pval_cutoff, dataset, datasets, return_all = F, merge_domain_data = F){
    res = mBenchmarkMotifs(datasets = dataset, 
                           descriptions = descriptions[datasets == dataset],
                           dir = "./", 
                           merge_domain_data = merge_domain_data, neg_set = neg_set,
                           domain_res_file = "./processed_data_files/what_we_find_VS_ELM_clust20171019.RData",
                           domain_results_obj = "res_count", motif_input_obj = "forSLIMFinder_Ready",
                           one_from_cloud = T, type = "QSLIMFinder",
                           dbfile_main = "./data_files/instances9606.gff",
                           dburl_main = "http://elm.eu.org/instances.gff?q=None&taxon=Homo%20sapiens&instance_logic=",
                           dbfile_query = "./data_files/instances10239.gff",
                           dburl_query = "http://elm.eu.org/instances.gff?q=all&taxon=irus&instance_logic=",
                           query_res_query_only = T, motif_types = motif_types,
                           all_res_excl_query = T,
                           seed = 21, N = N_negative_sets, replace = T, within1sequence = T,
                           query_predictor_col = "Sig", all_predictor_col = "Sig", normalise = normalise,
                           minoverlap = 2, maxgap = 0, motif_setup_months = motif_setup_month[datasets == dataset],
                           count_ranges_by = list(by = "IDs_domain_human", name = "motif_occ_per_domain",
                                                  normalise_by = "domain_count", normalised_name = "normalised_motif_occ_per_domain"),
                           filter_by_domain_data = filter_by_domain_data,
                           motif_pval_cutoff = motif_pval_cutoff)
    DT = data.table(N_query_prot_with_known_instances = res[[1]]$N_query_prot_with_known_instances,
               N_query_known_instances = res[[1]]$N_query_known_instances,
               N_query_prot_with_known_instances_found = res[[1]]$N_query_prot_with_known_instances_found,
               N_query_found_match_known_instances = res[[1]]$N_query_known_instances_found,
               N_query_total_instances_found = res[[1]]$N_query_total_instances_found,
               N_query_known_match_instances_found = res[[1]]$N_query_match_known_instances_found)
    if(return_all) list(res = res, DT = DT) else return(DT)
}
#####################################################################################
res_list_viral = myBenchmarkMotifs(neg_set = "all_instances",
                                   filter_by_domain_data = NULL,
                                   motif_pval_cutoff = 1,
                                   dataset = datasets[1], datasets)
sapply(colnames(benchmark), function(dataset) {
    prec05cutoff = 1 - min(benchmark["ROC_prediction",dataset][[1]]@cutoffs[[1]][
    benchmark["ROC_performance",dataset][[1]]@y.values[[1]] > 0.5], na.rm = T)
known_at_prec05cutoff = res_list_viral$qslimfinder.Full_IntAct3cloudfixF.FALSE$overlapping_GRanges_query[
    res_list_viral$qslimfinder.Full_IntAct3cloudfixF.FALSE$overlapping_GRanges_query$Sig <= prec05cutoff]
unique(paste0(seqnames(known_at_prec05cutoff), "=", known_at_prec05cutoff$ID))
})
#####################################################################################
plotVenn = function(res, ...){
    #grid.newpage()
    if(res$N_query_total_instances_found < res$N_query_known_instances) rotation.degree = 180 else rotation.degree = 0
    venn.plot <- VennDiagram::draw.pairwise.venn(
        area1 = res$N_query_total_instances_found,
        area2 = res$N_query_known_instances,
        cross.area = res$N_query_known_match_instances_found,
        category = c("discovered instances", "known instances"),
        fill = c("royalblue1", "seagreen1"),
        euler.d = F, scaled = F,
        lty = "blank",
        cat.pos = c(340, 175),
        cat.dist = c(0.05,0.05),
        ext.pos = 10,
        ext.dist = -0.05,
        ext.length = 0.8,
        ext.line.lwd = 2,
        ext.line.lty = "dashed",
        rotation.degree = rotation.degree,
        ind = F,
        ...
    );
    #grid.newpage()
    venn = gTree(children=venn.plot, name="venn")
    #grid.newpage()
    # since multiple matches are possible between datasets I need to edit numbers shown on a diagram
    venn = editGrob(venn, paste0(venn[["childrenOrder"]][7]),
                    label = paste0(res$N_query_found_match_known_instances,
                                   " / ",res$N_query_known_match_instances_found),
                    grep = T)
    #grid.newpage()
    if(rotation.degree == 0) discovered_ind = 5 else discovered_ind = 6
    venn = editGrob(venn, paste0(venn[["childrenOrder"]][discovered_ind]),
                    label = res$N_query_total_instances_found - res$N_query_found_match_known_instances,
                    grep = T)
    #grid.newpage()
    #grid.draw(venn)
    return(venn)
}
#####################################################################################
plotVenn(res_list_viral)
#####################################################################################
plotAnnotVenn = function(res, main = "", sub = "",
                         main_fontsize = 22, sub_fontsize = 17,
                         main_lineheight = 1.4, sub_lineheight = 0.9,
                         cex = 1.8, cat.cex = 1.8,
                         fontfamily = "Arial", ...) { # "serif" "Arial"
    
    lay = matrix(c(1), nrow = 1, ncol = 1)
    annotated_venn = plotVenn(res, fontfamily = fontfamily, cat.fontfamily = fontfamily,
                              cex = cex, cat.cex = cat.cex)
    if(sub != "") {
        lay = matrix(c(1,2,2,2,2,2), nrow = 6, ncol = 1)
        sub_grob = textGrob(sub,
                            gp=gpar(fontsize=sub_fontsize, fontfamily = fontfamily,
                                    lineheight = sub_lineheight, ...))
        annotated_venn = gridExtra::arrangeGrob(sub_grob,
                                                annotated_venn, ncol=1,
                                                layout_matrix = lay)
    }
    
    if(main != ""){
        lay = matrix(c(1,2,2,2,2,2), nrow = 6, ncol = 1)
        main_grob = textGrob(main,
                             gp=gpar(fontsize=main_fontsize, fontfamily = fontfamily,
                                     lineheight = main_lineheight, ...))
        annotated_venn = gridExtra::arrangeGrob(main_grob,
                                                annotated_venn, ncol=1,
                                                layout_matrix = lay)
    } 
    
    #grid.newpage()
    #grid.draw(annotated_venn)
    return(annotated_venn)
}
#####################################################################################
plotAnnotVenn(res_list_viral, main = "corrected p-value < 0.3")
#####################################################################################
mPlotAnnotVenn = function(filter_by_domain_data = NULL,
                          motif_pval_cutoffs = c(0.3, 0.05, 0.002),
                          motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                          dataset = datasets[1], datasets = datasets, fontfamily = "Arial") {
        annotated_venns = lapply(motif_pval_cutoffs, function(motif_pval_cutoff){
            res_instances = myBenchmarkMotifs(neg_set = "all_instances",
                                              filter_by_domain_data = filter_by_domain_data,
                                              motif_pval_cutoff = motif_pval_cutoff,
                                              dataset = dataset, datasets)
            res_proteins = myBenchmarkMotifs(neg_set = "all_proteins",
                                             filter_by_domain_data = filter_by_domain_data,
                                             motif_pval_cutoff = motif_pval_cutoff,
                                             dataset = dataset, datasets)
            annotated_venn = plotAnnotVenn(res_instances,
                                           main = paste0(motif_pval_tags[motif_pval_cutoffs == motif_pval_cutoff],
                                                         ": adj. p-value < ", signif(motif_pval_cutoff, 2)),
                                           sub = paste0(res_instances$N_query_prot_with_known_instances,
                                                        " proteins with known SLIMs (",
                                                        res_instances$N_query_prot_with_known_instances_found," proteins recovered)\n",
                                                        res_proteins$N_query_total_instances_found," SLIMs predicted in other proteins"),
                                           fontfamily = fontfamily)
        })
        gridExtra::arrangeGrob(grobs = annotated_venns, ncol=1)
}
#####################################################################################
figureVenn = function(filter_by_domain_data = NULL,
                      main = "",
                      motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                      datasets = datasets,
                      stringent_prec = 0.5,
                      lenient_cutoff = 0.3,
                      motif_pval_cutoffs_IntAct = NULL,
                      motif_pval_cutoffs_Vidal = NULL,
                      motif_pval_cutoffs_viral = NULL,
                      fontfamily = "Arial", ...){
    ################################################################################ IntAct
    res_list_IntAct = myBenchmarkMotifs(neg_set = "all_instances",
                                        filter_by_domain_data = filter_by_domain_data,
                                        motif_pval_cutoff = 1,
                                        dataset = datasets[1], datasets, return_all = T)
    rocr_list_IntAct = mBenchmarkMotifsROC(res_list_IntAct$res, measure1 = "prec", measure2 = "rec", ROCR_data = T)
    if(is.null(motif_pval_cutoffs_IntAct)){
        motif_pval_cutoffs_IntAct = c(lenient_cutoff,
                                      1 - min(rocr_list_IntAct[1,1][[1]]@cutoffs[[1]][rocr_list_IntAct[2,1][[1]]@y.values[[1]] >
                                                                                          rocr_list_IntAct[2,1][[1]]@x.values[[1]]], na.rm = T),
                                      1 - min(rocr_list_IntAct[1,1][[1]]@cutoffs[[1]][rocr_list_IntAct[2,1][[1]]@y.values[[1]] > stringent_prec],
                                              na.rm = T))
    }
    big_plot_IntAct = mPlotAnnotVenn(filter_by_domain_data = filter_by_domain_data,
                                     motif_pval_cutoffs = motif_pval_cutoffs_IntAct,
                                     motif_pval_tags = motif_pval_tags,
                                     dataset = datasets[1], datasets = datasets, fontfamily = fontfamily)
    IntAct_grob = textGrob("including all human data (IntAct)",
                           gp=gpar(fontsize=25, fontfamily = fontfamily,
                                   lineheight = 1.4, ...))
    ################################################################################ Vidal
    res_list_Vidal = myBenchmarkMotifs(neg_set = "all_instances",
                                        filter_by_domain_data = filter_by_domain_data,
                                        motif_pval_cutoff = 1,
                                        dataset = datasets[2], datasets, return_all = T)
    rocr_list_Vidal = mBenchmarkMotifsROC(res_list_Vidal$res, measure1 = "prec", measure2 = "rec", ROCR_data = T)
    if(is.null(motif_pval_cutoffs_Vidal)){
        motif_pval_cutoffs_Vidal = c(lenient_cutoff,
                                      1 - min(rocr_list_Vidal[1,1][[1]]@cutoffs[[1]][rocr_list_Vidal[2,1][[1]]@y.values[[1]] >
                                                                                          rocr_list_Vidal[2,1][[1]]@x.values[[1]]], na.rm = T),
                                      1 - min(rocr_list_Vidal[1,1][[1]]@cutoffs[[1]][rocr_list_Vidal[2,1][[1]]@y.values[[1]] > stringent_prec],
                                              na.rm = T))
    }
    big_plot_Vidal = mPlotAnnotVenn(filter_by_domain_data = filter_by_domain_data,
                                     motif_pval_cutoffs = motif_pval_cutoffs_Vidal,
                                     motif_pval_tags = motif_pval_tags,
                                     dataset = datasets[2], datasets = datasets, fontfamily = fontfamily)
    Vidal_grob = textGrob("including human data (Vidal)",
                           gp=gpar(fontsize=25, fontfamily = fontfamily,
                                   lineheight = 1.4, ...))
    ################################################################################ viral only
    res_list_viral = myBenchmarkMotifs(neg_set = "all_instances",
                                       filter_by_domain_data = filter_by_domain_data,
                                       motif_pval_cutoff = 1,
                                       dataset = datasets[3], datasets, return_all = T)
    rocr_list_viral = mBenchmarkMotifsROC(res_list_viral$res, measure1 = "prec", measure2 = "rec", ROCR_data = T)
    if(is.null(motif_pval_cutoffs_viral)){
        motif_pval_cutoffs_viral = c(lenient_cutoff,
                                     1 - min(rocr_list_viral[1,1][[1]]@cutoffs[[1]][rocr_list_viral[2,1][[1]]@y.values[[1]] >
                                                                                        rocr_list_viral[2,1][[1]]@x.values[[1]]], na.rm = T),
                                     1 - min(rocr_list_viral[1,1][[1]]@cutoffs[[1]][rocr_list_viral[2,1][[1]]@y.values[[1]] > stringent_prec],
                                             na.rm = T))
    } 
    
    big_plot_viral = mPlotAnnotVenn(filter_by_domain_data = filter_by_domain_data,
                                    motif_pval_cutoffs = motif_pval_cutoffs_viral,
                                    motif_pval_tags = motif_pval_tags,
                                    dataset = datasets[3], datasets = datasets, fontfamily = fontfamily)
    viral_grob = textGrob("only viral-human data",
                          gp=gpar(fontsize=25, fontfamily = fontfamily,
                                  lineheight = 1.4, ...))
    
    big_plot = gridExtra::arrangeGrob(big_plot_IntAct, big_plot_Vidal, big_plot_viral, nrow=1)
    big_text = gridExtra::arrangeGrob(IntAct_grob, Vidal_grob, viral_grob, nrow=1)
    big_plot_text = gridExtra::arrangeGrob(big_text, big_plot, ncol=1,
                                           layout_matrix = matrix(c(1,rep(2,6)),
                                                                  nrow = 7, ncol = 1))
    if(main != ""){
        main_grob = textGrob(main,
                          gp=gpar(fontsize=27, fontfamily = fontfamily,
                                  lineheight = 1, ...))
        big_plot_text = gridExtra::arrangeGrob(main_grob, big_plot_text, ncol=1,
                                           layout_matrix = matrix(c(1,rep(2,7)),
                                                                  nrow = 8, ncol = 1))
    }
    
    list(big_plot_text = big_plot_text,
         res_list_IntAct = res_list_IntAct, rocr_list_IntAct = rocr_list_IntAct, 
         res_list_Vidal = res_list_Vidal, rocr_list_Vidal = rocr_list_Vidal,
         res_list_viral = res_list_viral, rocr_list_viral = rocr_list_viral, 
         big_plot_IntAct = big_plot_IntAct, IntAct_grob = IntAct_grob,
         big_plot_Vidal = big_plot_Vidal, Vidal_grob = Vidal_grob,
         big_plot_viral = big_plot_viral, viral_grob = viral_grob)
}
#####################################################################################
replotFigureVenn = function(figure, main, fontfamily = "Arial", ...){
    IntAct_grob = textGrob("including all human data (IntAct)",
                           gp=gpar(fontsize=24, fontfamily = fontfamily,
                                   lineheight = 1, ...))
    Vidal_grob = textGrob("including human data (Vidal)",
                           gp=gpar(fontsize=24, fontfamily = fontfamily,
                                   lineheight = 1, ...))
    viral_grob = textGrob("only viral-human data",
                          gp=gpar(fontsize=24, fontfamily = fontfamily,
                                  lineheight = 1, ...))
    
    big_plot = gridExtra::arrangeGrob(rectGrob(gp=gpar(col = "white")),
                                      figure$big_plot_viral,
                                      rectGrob(gp=gpar(col = "white")),
                                      figure$big_plot_IntAct,
                                      rectGrob(gp=gpar(col = "white")),
                                      figure$big_plot_Vidal, nrow=1,
                                      layout_matrix = matrix(c(1,rep(2,6),
                                                               3,rep(4,6),
                                                               5,rep(6,6)),
                                                                  nrow = 1, ncol = 21))
    big_text = gridExtra::arrangeGrob(viral_grob, IntAct_grob, Vidal_grob, nrow=1)
    big_plot_text = gridExtra::arrangeGrob(big_text, big_plot, ncol=1,
                                           layout_matrix = matrix(c(1,rep(2,7)),
                                                                  nrow = 8, ncol = 1))
    if(main != ""){
        main_grob = textGrob(main,
                          gp=gpar(fontsize=27, fontfamily = fontfamily,
                                  lineheight = 1, ...))
        big_plot_text = gridExtra::arrangeGrob(main_grob, big_plot_text, ncol=1,
                                           layout_matrix = matrix(c(1,rep(2,7)),
                                                                  nrow = 8, ncol = 1))
    }
    list(big_plot_text = big_plot_text,
         res_list_IntAct = figure$res_list_IntAct, rocr_list_IntAct = figure$rocr_list_IntAct, 
         res_list_Vidal = figure$res_list_Vidal, rocr_list_Vidal = figure$rocr_list_Vidal,
         res_list_viral = figure$res_list_viral, rocr_list_viral = figure$rocr_list_viral, 
         big_plot_IntAct = figure$big_plot_IntAct, IntAct_grob = IntAct_grob,
         big_plot_Vidal = figure$big_plot_Vidal, Vidal_grob = Vidal_grob,
         big_plot_viral = figure$big_plot_viral, viral_grob = viral_grob)
}
##################################################################################### "MOD", "LIG", "DOC"

figureMODLIGDOC = figureVenn(filter_by_domain_data = NULL,
                    main = "human network searched for motifs present in viral proteins (not filtered by domain)",
                    motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                    datasets = datasets,
                    motif_pval_cutoffs_viral = c(0.3, 0.1, 0.03))
figureMODLIGDOC2 = replotFigureVenn(figureMODLIGDOC,
                 main = "")
ggsave(filename = "Venn_not_filtered_MOD_LIG_DOC.svg", plot = figureMODLIGDOC2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure2$big_plot_text)

figureMODLIGDOC2_filtered = figureVenn(filter_by_domain_data = "p.value < 0.5",
                             main = "human network searched for motifs present in viral proteins (filtered by domain)",
                             motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                             datasets = datasets)
figureMODLIGDOC2_filtered2 = replotFigureVenn(figureMODLIGDOC2_filtered,
                 main = "")
ggsave(filename = "Venn_filtered_MOD_LIG_DOC.svg", plot = figureMODLIGDOC2_filtered2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure_filtered2$big_plot_text)
##################################################################################### ^ copy optimal and stringent motif_pval_cutoffs
##################################################################################### "MOD", "LIG", "DOC", "DEG", "CLV"
motif_types = c("MOD", "LIG", "DOC", "DEG", "CLV")

figureMODLIGDOC2DEGCLV = figureVenn(filter_by_domain_data = NULL,
                    main = "human network searched for motifs present in viral proteins (not filtered by domain)",
                    motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                    datasets = datasets,
                    motif_pval_cutoffs_IntAct = c(0.3, 0.063, 0.003),
                    motif_pval_cutoffs_Vidal = c(0.3, 0.069, 8e-05),
                    motif_pval_cutoffs_viral = c(0.3, 0.1, 0.03))
figureMODLIGDOC2DEGCLV2 = replotFigureVenn(figureMODLIGDOC2DEGCLV,
                 main = "")
ggsave(filename = "Venn_not_filtered_MOD_LIG_DOC_DEG_CLV.svg", plot = figureMODLIGDOC2DEGCLV2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure2$big_plot_text)

figureMODLIGDOC2DEGCLV_filtered = figureVenn(filter_by_domain_data = "p.value < 0.5",
                             main = "human network searched for motifs present in viral proteins (filtered by domain)",
                             motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                             datasets = datasets,
                    motif_pval_cutoffs_IntAct = c(0.3, 0.039, 0.002),
                    motif_pval_cutoffs_Vidal = c(0.3, 0.027, 7.4e-05),
                    motif_pval_cutoffs_viral = c(0.3, 0.14, 0.04))
figureMODLIGDOC2DEGCLV_filtered2 = replotFigureVenn(figureMODLIGDOC2DEGCLV_filtered,
                 main = "")
ggsave(filename = "Venn_filtered_MOD_LIG_DOC_DEG_CLV.svg", plot = figureMODLIGDOC2DEGCLV_filtered2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure_filtered2$big_plot_text)
##################################################################################### "MOD", "LIG", "DOC" "DEG", "CLV", "TRG"
motif_types = c("MOD", "LIG", "DOC", "DEG", "CLV", "TRG")

figureMODLIGDOC2DEGCLVTRG = figureVenn(filter_by_domain_data = NULL,
                    main = "human network searched for motifs present in viral proteins (not filtered by domain)",
                    motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                    datasets = datasets,
                    motif_pval_cutoffs_IntAct = c(0.3, 0.063, 0.003),
                    motif_pval_cutoffs_Vidal = c(0.3, 0.069, 8e-05),
                    motif_pval_cutoffs_viral = c(0.3, 0.1, 0.03))
figureMODLIGDOC2DEGCLVTRG2 = replotFigureVenn(figureMODLIGDOC2DEGCLVTRG,
                 main = "")
ggsave(filename = "Venn_not_filtered_MOD_LIG_DOC_DEG_CLV_TRG.svg", plot = figureMODLIGDOC2DEGCLVTRG2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure2$big_plot_text)

figureMODLIGDOC2DEGCLVTRG_filtered = figureVenn(filter_by_domain_data = "p.value < 0.5",
                             main = "human network searched for motifs present in viral proteins (filtered by domain)",
                             motif_pval_tags = c("lenient cutoff", "optimal cutoff", "stringent cutoff"),
                             datasets = datasets,
                    motif_pval_cutoffs_IntAct = c(0.3, 0.039, 0.002),
                    motif_pval_cutoffs_Vidal = c(0.3, 0.027, 7.4e-05),
                    motif_pval_cutoffs_viral = c(0.3, 0.14, 0.04))
figureMODLIGDOC2DEGCLVTRG_filtered2 = replotFigureVenn(figureMODLIGDOC2DEGCLVTRG_filtered,
                 main = "")
ggsave(filename = "Venn_filtered_MOD_LIG_DOC_DEG_CLV_TRG.svg", plot = figureMODLIGDOC2DEGCLVTRG_filtered2$big_plot_text,
       device = "svg", width = 19.3, height = 11.60638, units = "in")
#grid.newpage()
#grid.draw(figure_filtered2$big_plot_text)
#####################################################################################
```

# protein-domain-motif-protein network

```{r}
motif_types = c("MOD", "LIG", "DOC")
# get motif & domains (stringent threshold)
IntAct = myBenchmarkMotifs(neg_set = "all_proteins",
                  filter_by_domain_data = "p.value < 0.5",
                  motif_pval_cutoff = 0.002,
                  dataset = datasets[1], datasets,
                  return_all = T, merge_domain_data = T)
IntAct_lenient = myBenchmarkMotifs(neg_set = "all_proteins",
                  filter_by_domain_data = "p.value < 0.5",
                  motif_pval_cutoff = 0.3,
                  dataset = datasets[1], datasets,
                  return_all = T, merge_domain_data = T)
viral = myBenchmarkMotifs(neg_set = "all_proteins",
                  filter_by_domain_data = "p.value < 0.5",
                  motif_pval_cutoff = 0.04,
                  dataset = datasets[3], datasets,
                  return_all = T, merge_domain_data = T)
```

# how many motif instance at a stringent threshold after domain filtering?

```{r}
instances = paste0(IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$query, "_",
                         IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$Pattern)
instances = instances[!is.na(IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$IDs_domain_human)]
instances = unique(instances)
proteins = sapply((strsplit(instances, "_")), function(x) x[1])
patterns = sapply((strsplit(instances, "_")), function(x) x[2])
NLS_ind = grepl("KR", patterns) &
    !grepl("G", patterns) & !grepl("P", patterns) & !grepl("H", patterns) &
    !grepl("Q", patterns) & !grepl("S", patterns) & !grepl("E", patterns)

non_NLS_instances = instances[!NLS_ind]
non_NLS_proteins = unique(proteins[!NLS_ind])

# how many proteins with motifs?
length(non_NLS_proteins)
# - 6 known
length(non_NLS_proteins) - 6 
```

# do results overlap?

```{r}
IntAct_known = unique(paste0(seqnames(IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query), "_",
                             start(IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query), "_",
                             end(IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query)))
viral_known = unique(paste0(seqnames(viral$res$qslimfinder.all_viral_interaction3.FALSE$overlapping_GRanges_query), "_",
                            start(viral$res$qslimfinder.all_viral_interaction3.FALSE$overlapping_GRanges_query), "_",
                            end(viral$res$qslimfinder.all_viral_interaction3.FALSE$overlapping_GRanges_query)))
union_known = unique(c(IntAct_known, viral_known))

IntAct_only = union_known %in% IntAct_known
viral_only = union_known %in% viral_known

names(IntAct_only) = union_known
names(viral_only) = union_known

# human network missed
viral_net[unique_position == "P03129_21_26"]
viral_net[unique_position == "P03129_32_36"]
viral_net[unique_position == "P03255_121_126"]

# viral network missed
IntAct_net[unique_position == "P03428_736_739"]
IntAct_net[unique_position == "P03428_752_755"]
IntAct_net[unique_position == "P06463_148_155"]
```

```{r}
# convert motif GRanges to simple network table
bechmark2Net = function(occurence_query = IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query,
                        matching_known = IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query,
                        top_domain = T) {
    Net = data.table(Viral_protein = occurence_query$query,
                     Motif_Sig = occurence_query$Sig,
                     Domain_p.value = occurence_query$p.value,
                     Pattern = occurence_query$Pattern,
                     Human_recognition_domain = occurence_query$IDs_domain_human,
                     Human_protein = occurence_query$interacts_with,
                     unique_position = paste0(GenomicRanges::seqnames(occurence_query), "_",
                                              GenomicRanges::start(occurence_query), "_",
                                              GenomicRanges::end(occurence_query)))
    Net = unique(Net)
    if(top_domain){
        Net[, top_domain := Domain_p.value == min(Domain_p.value), by = unique_position]
        Net = Net[top_domain == TRUE]
        Net[, top_domain := NULL]
    }
    
    matching_known = data.table(unique_position = paste0(GenomicRanges::seqnames(matching_known), "_",
                                                         GenomicRanges::start(matching_known), "_",
                                                         GenomicRanges::end(matching_known)),
                                matching_known = "matching_known")
    Net = merge(Net, matching_known, by = "unique_position", all.x = T, all.y = F, allow.cartesian=TRUE)
    unique(Net)
}

IntAct_net = bechmark2Net(occurence_query = IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query,
                        matching_known = IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query)
viral_net = bechmark2Net(occurence_query = viral$res$qslimfinder.all_viral_interaction3.FALSE$occurence_query,
                        matching_known = viral$res$qslimfinder.all_viral_interaction3.FALSE$overlapping_GRanges_query)
```

### find ELM domains

```{r ELM_domains}
if(!file.exists("./data_files/interactiondomains.tsv")) downloader::download("http://elm.eu.org/interactiondomains.tsv","./data_files/interactiondomains.tsv")
interactiondomains = fread("./data_files/interactiondomains.tsv")
interactiondomains[, pfam_id := `Interaction Domain Id`]

domains_known = interactiondomains[, unique(pfam_id)]

InterProScan_domains = readInterProGFF3("./processed_data_files/all_human_viral_protein_domains.gff3.gz")
# get InterProID to member database ID mapping
InterPro2memberDB = getInterPro2memberDB(InterProScan_domains)
InterPro2memberDB = InterPro2memberDB[complete.cases(InterPro2memberDB)]
domains_known_mapped = unique(InterPro2memberDB[memberDBID %in% domains_known | InterProID %in% domains_known, InterProID])

IntAct_all_dom = myBenchmarkMotifs(neg_set = "all_proteins",
                  filter_by_domain_data = NULL,
                  motif_pval_cutoff = 0.002,
                  dataset = datasets[1], datasets,
                  return_all = T, merge_domain_data = T)
dom = IntAct_all_dom$res$qslimfinder.Full_IntAct3.FALSE$domain_res$data_with_pval
dom[IDs_domain_human %in% domains_known_mapped, SLIM_binding := "yes"]
dom[!IDs_domain_human %in% domains_known_mapped, SLIM_binding := "no / not known"]
# how many human proteins have SLIM-binding domains
dom[SLIM_binding == "yes", uniqueN(IDs_interactor_human)]
# how many viral proteins target human proteins that have SLIM-binding domains
dom[SLIM_binding == "yes", uniqueN(IDs_interactor_viral)]

dom = unique(dom[,.(IDs_interactor_viral, IDs_domain_human, p.value, domain_type, SLIM_binding)])
ggplot(dom, aes(p.value, color = SLIM_binding, fill = SLIM_binding)) +
    geom_density(alpha = 0.3) + theme_light()
ks.test(dom[SLIM_binding == "no / not known", p.value],
        dom[SLIM_binding == "yes", p.value], 
        alternative = "less")
dom05 = dom[p.value < 0.5]
dom05[, uniqueN(IDs_interactor_viral)]
dom05[, uniqueN(IDs_domain_human)]

# how many motif per protein
table(IntAct_all_dom$res$qslimfinder.Full_IntAct3.FALSE$instances_query$ID, IntAct_all_dom$res$qslimfinder.Full_IntAct3.FALSE$instances_query$Primary_Acc)[, table(IntAct_all_dom$res$qslimfinder.Full_IntAct3.FALSE$instances_query$Primary_Acc) > 2]
```

```{r, eval=FALSE}
# get readable names for proteins and domains 
readableNet = function(Net){
    proteins2gene_names = lapply(unique(c(Net$Viral_protein, Net$Human_protein)), function(uniprot_id) {
        suppressMessages({
            proteins2gene_names = fread(paste0("https://www.uniprot.org/uniprot/?query=id:", uniprot_id,
                                               "&format=tab&columns=id,genes,entry%20name,protein%20names,organism,organism-id"),
                                        stringsAsFactors = F)
        })
        proteins2gene_names
    })
    proteins2gene_names = Reduce(rbind, proteins2gene_names)
    setnames(proteins2gene_names, colnames(proteins2gene_names),
             c( "Entry", "Gene_names", "Entry_name", "Protein_names", "Organism", "Organism_ID"))
    proteins2gene_names[, Gene_names := gsub(" .*", "", Gene_names)] # pick only first gene name
    proteins2gene_names[, Protein_names := gsub(" \\(.*", "", Protein_names)] # pick only first protein name
    # fwrite(proteins2gene_names, "./results/IntAct_network_proteins2gene_names", sep = "\t")
    proteins2gene_names_human = proteins2gene_names[Organism_ID == 9606]
    proteins2gene_names_viral = proteins2gene_names[Organism_ID != 9606]
    
    proteins2gene_names_human[, c("Organism", "Organism_ID") := NULL]
    proteins2gene_names_viral[, c("Organism", "Organism_ID") := NULL]
    
    setnames(proteins2gene_names_human, colnames(proteins2gene_names_human),
             c("Human_protein", "Human_gene_names", "Human_entry_name", "Human_protein_names"))
    setnames(proteins2gene_names_viral, colnames(proteins2gene_names_viral),
             c("Viral_protein", "Viral_gene_names", "Viral_entry_name", "Viral_protein_names"))
    
    Net = merge(Net, proteins2gene_names_human, by = "Human_protein", all.x = T, all.y = F)
    Net = merge(Net, proteins2gene_names_viral, by = "Viral_protein", all.x = T, all.y = F)
    
    InterProEntryTypes = getInterProEntryTypes("./data_files/entry.list")
    InterProEntryTypes[, ENTRY_TYPE := NULL]
    setnames(InterProEntryTypes, colnames(InterProEntryTypes),
             c("Human_recognition_domain", "Human_recognition_domain_name"))
    Net = merge(Net, InterProEntryTypes, by = "Human_recognition_domain", all.x = T, all.y = F)
    
    # mark known SLIM-interacting domains
    Net[, in_ELM := Human_recognition_domain %in% domains_known_mapped]
    
    # reorder column names 
    Net = Net[, .(Pattern, Human_recognition_domain_name, Domain_p.value, unique_position,
                  Human_protein, in_ELM, matching_known,
                  Viral_protein_names, Viral_entry_name, Human_protein_names, Human_entry_name,
                  Human_recognition_domain, Motif_Sig, Viral_protein,
                  Viral_gene_names, Human_gene_names)]
    setorder(Net, Viral_protein, Motif_Sig, Domain_p.value)
    unique(Net)
}

IntAct_net = readableNet(IntAct_net)
viral_net = readableNet(viral_net)

fwrite(IntAct_net, "./results/IntAct_net_motif_domain_readable.tsv", sep = "\t")
fwrite(viral_net, "./results/viral_net_motif_domain_readable.tsv", sep = "\t")

fwrite(unique(IntAct_net[,.(Pattern)]), "./results/IntAct_net_patterns.tsv", sep = "\t")
fwrite(unique(viral_net[,.(Pattern)]), "./results/viral_net_patterns.tsv", sep = "\t")
```

Armadillo like domains:  
Viral proteins `r IntAct_net[grepl("Armadillo", Human_recognition_domain_name) & grepl("KR", Pattern), uniqueN(Viral_protein)]`.  
Viral motifs `r IntAct_net[grepl("Armadillo", Human_recognition_domain_name) & grepl("KR", Pattern), uniqueN(unique_position)]`.  
Human proteins `r IntAct_net[grepl("Armadillo", Human_recognition_domain_name) & grepl("KR", Pattern), uniqueN(Human_protein)]`.   

PDZ domains:  
Viral proteins `r IntAct_net[grepl("PDZ", Human_recognition_domain_name) & grepl("\\$", Pattern), uniqueN(Viral_protein)]`.  
Viral motifs `r IntAct_net[grepl("PDZ", Human_recognition_domain_name) & grepl("\\$", Pattern), uniqueN(unique_position)]`.  
Human proteins `r IntAct_net[grepl("PDZ", Human_recognition_domain_name) & grepl("\\$", Pattern), uniqueN(Human_protein)]`.   

PDZ motifs:  
Viral proteins `r IntAct_net[grepl("\\$", Pattern), uniqueN(Viral_protein)]`.  
Viral motifs `r IntAct_net[grepl("\\$", Pattern), uniqueN(unique_position)]`.  
Human proteins `r IntAct_net[grepl("\\$", Pattern), uniqueN(Human_protein)]`.   


```{r more_motif_summaries}
motifSummary = function(benchmarkMotifsResult = IntAct$res$qslimfinder.Full_IntAct3.FALSE,
                        readable_net = IntAct_net,
                        motif = "E.V..G.{0,2}N.{0,1}Q",
                        viral_query = NULL,
                        human_seed = NULL,
                        results_dir = "./results/"){
    if(is.null(viral_query)) viral_query = readable_net$Viral_protein
    if(is.null(human_seed)) human_seed = readable_net$Human_protein
    if(is.null(motif)) motif = readable_net$Pattern
    
    readable_net_ind = readable_net$Pattern %in% motif & 
        readable_net$Viral_protein %in% viral_query & 
        readable_net$Human_protein %in% human_seed
    occurence_query_ind = benchmarkMotifsResult$occurence_query$Pattern %in% motif &
        benchmarkMotifsResult$occurence_query$query %in% viral_query &
        benchmarkMotifsResult$occurence_query$interacts_with %in% human_seed
    
    occurence_query = benchmarkMotifsResult$occurence_query[occurence_query_ind]
    occurence_query = data.table(UP = occurence_query$UP, UPNum = occurence_query$UPNum,
                                 unique_position = paste0(GenomicRanges::seqnames(occurence_query), "_",
                                                     GenomicRanges::start(occurence_query), "_",
                                                     GenomicRanges::end(occurence_query)),
                                 Occ = occurence_query$Occ, SeqNum = occurence_query$SeqNum, 
                                 domain_count_per_IDs_interactor_viral = occurence_query$domain_count_per_IDs_interactor_viral,
                                 IDs_interactor_viral_degree = occurence_query$IDs_interactor_viral_degree,
                                 IDs_domain_human = occurence_query$IDs_domain_human,
                                 Human_protein = occurence_query$interacts_with)
    occurence_query = unique(occurence_query)
    
    occurence_ind = benchmarkMotifsResult$occurence$Pattern %in% motif &
        benchmarkMotifsResult$occurence$query %in% viral_query &
        benchmarkMotifsResult$occurence$interacts_with %in% human_seed
    occurence_seqnames = as.character(GenomicRanges::seqnames(benchmarkMotifsResult$occurence[occurence_ind]))
    occurence_seqnames = unique(occurence_seqnames)
    
    summary = paste0(
        "viral query protein with motif: ",
        paste0(unique(paste0(readable_net[readable_net_ind]$unique_position, " = ",
                             readable_net[readable_net_ind]$Pattern, " = ",
                             readable_net[readable_net_ind]$Viral_protein_names, " (",
                             readable_net[readable_net_ind]$Viral_entry_name, ") ")), collapse = " | "), "\n",
        uniqueN(paste0(readable_net[readable_net_ind]$unique_position, "=", readable_net[readable_net_ind]$Viral_protein_names)),
        "\n\n",
        "viral targeted human proteins: ",
        paste0(unique(paste0(readable_net[readable_net_ind]$Human_protein, " = ", readable_net[readable_net_ind]$Human_protein_names)), collapse = " | "), "\n",
        uniqueN(paste0(readable_net[readable_net_ind]$Human_protein, "=", readable_net[readable_net_ind]$Human_protein_names)),
        "\n\n",
        "UP support / UP: ",
        paste0(unique(paste0("(", occurence_query$unique_position," in ", occurence_query$Human_protein," dataset) ",
            occurence_query$UP, "/",
            occurence_query$UPNum)), collapse = " | "),
        "\n\n",
        "Occurence / SeqNum: ",
        paste0(unique(paste0("(", occurence_query$unique_position," in ", occurence_query$Human_protein," dataset) ",
            occurence_query$Occ, "/",
            occurence_query$SeqNum)), collapse = " | "),
        "\n\n",
        "top-1 domain(s): ",
        paste0(paste0("(", occurence_query$unique_position,") ", occurence_query$IDs_domain_human), collapse = " | "),
        "\n\n",
        "domain_count_per_IDs_interactor_viral / IDs_interactor_viral_degree: ",
        paste0(paste0(
            occurence_query$domain_count_per_IDs_interactor_viral, "/",
            occurence_query$IDs_interactor_viral_degree), collapse = " | "),
        "\n\n",
        "all with motif: ",
        paste0(occurence_seqnames, collapse = " | "), "\n",
        "N ", uniqueN(occurence_seqnames),
        "\nN non-query with motif: ", uniqueN(occurence_seqnames) -
            uniqueN(readable_net[readable_net_ind]$Viral_protein),
        "\n\n"
    )
    
    readable_net2 = readable_net[readable_net_ind]
    # For Cytoscape: network table
    Nrows = nrow(readable_net2)
    readable_net3 = data.table(source = c(readable_net2$Human_protein,   readable_net2$Human_recognition_domain, readable_net2$Pattern),
                                target = c(readable_net2$Human_recognition_domain, readable_net2$Pattern,          readable_net2$Viral_protein),
                                source_readable = c(paste0(readable_net2$Human_protein_names, " (", readable_net2$Human_gene_names, ")"),  
                                                    readable_net2$Human_recognition_domain_name,
                                                    readable_net2$Pattern),
                                target_readable = c(readable_net2$Human_recognition_domain_name,
                                                    readable_net2$Pattern,
                                                    paste0(readable_net2$Viral_protein_names, " (", readable_net2$Viral_entry_name, ")")),
                                #Motif_Sig =    c(rep(1, Nrows),                   rep(1, Nrows),              readable_net2$Motif_Sig),
                                type_source = c(rep("human_protein", Nrows), rep("human_domain", Nrows),  rep("viral_motif", Nrows)),
                                type_target = c(rep("human_domain", Nrows), rep("viral_motif", Nrows),  rep("viral_protein", Nrows)))
    # For Cytoscape: node table
    network_types = unique(data.table(node = c(readable_net3$source, readable_net3$target),
                                             node_readable = c(readable_net3$source_readable, readable_net3$target_readable),
                                             type = c(readable_net3$type_source, readable_net3$type_target),
                                             scale_by = c(c(rep("", Nrows), readable_net2$Domain_p.value, readable_net2$Motif_Sig,
                                                            readable_net2$Domain_p.value, readable_net2$Motif_Sig, rep("", Nrows)))))
    readable_net3_file = paste0(as.character(as.list(match.call())[["readable_net"]]), "_",
                                paste0(unique(readable_net[readable_net_ind]$unique_position), collapse = "_"), "_", 
                                paste0(unique(readable_net[readable_net_ind]$Human_protein), collapse = "_"))
    if(nchar(readable_net3_file) > 100) readable_net3_file = digest::digest(readable_net3_file, "sha1")
    readable_net3_file = paste0(results_dir, readable_net3_file)
    readable_net3 = unique(readable_net3[complete.cases(readable_net3)])
    fwrite(readable_net3, 
           readable_net3_file,
           sep = "\t")
    fwrite(network_types,
           paste0(readable_net3_file,"types"),
           sep = "\t")
    
    cat(summary)
}
```

```{r view_summaries}
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "E.V..G.{0,2}N.{0,1}Q", viral_query = "B4URF7", human_seed = NULL)
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "E.V..G.{0,2}N.{0,1}Q", viral_query = "C5E527", human_seed = NULL)

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "[ST].[ILV]$", viral_query = NULL, human_seed = NULL)

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "DSG", viral_query = "P03070", human_seed = NULL)
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "DSG", viral_query = "P05919", human_seed = NULL)
IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query[IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$Pattern == "DSG"]

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "EI[IV]QQ", viral_query = NULL, human_seed = NULL)
 
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = NULL, viral_query = "P03126", human_seed = NULL)

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "LR.{0,2}G.G.T", viral_query = NULL, human_seed = NULL)
IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query[IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$Pattern == "LR.{0,2}G.G.T"]

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "PE.AEA", viral_query = NULL, human_seed = NULL)
IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query[IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$Pattern == "PE.AEA"]


motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "P..P.[HKR]", viral_query = NULL, human_seed = NULL)
IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query[IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query$Pattern == "PE.AEA"]
```

```{r write_cytoscape_files}
IntAct_net = bechmark2Net(occurence_query = IntAct$res$qslimfinder.Full_IntAct3.FALSE$occurence_query,
                        matching_known = IntAct$res$qslimfinder.Full_IntAct3.FALSE$overlapping_GRanges_query,
                        top_domain = F)
IntAct_net = readableNet(IntAct_net)

# all motifs and domains
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = NULL, viral_query = NULL, human_seed = NULL)

# known PDZ motifs
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = NULL, viral_query = "P03126", human_seed = NULL)

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = c("ET..$","[ST].V$", "[DE]T.[ILV]$"), viral_query = "P06463", human_seed = NULL)

# Candidate PDZ motif
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = NULL, viral_query = "Q2PJP0", human_seed = NULL)
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = c("[DE]..V$", "[ST].[ILV]$", "[ST].V$"), viral_query = "Q2PJP0", human_seed = NULL)

motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = NULL, viral_query = "P50804", human_seed = NULL)

# candidate SH3 domain motif
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "P..P.[HKR]", viral_query = NULL, human_seed = NULL)

# candidate WD40 domain motif
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "DSG", viral_query = NULL, human_seed = NULL)
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "E.V..G.{0,2}N.{0,1}Q", viral_query = NULL, human_seed = NULL)

# candidate LR.{0,2}G.G.T motif recognised by double-stranded RNA binding domain
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "LR.{0,2}G.G.T", viral_query = NULL, human_seed = NULL)
# candidate EI[IV]QQ motif potentially recognised by the EF hand domain
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "EI[IV]QQ", viral_query = NULL, human_seed = NULL)

# candidate motif L.{0,1}Q.LR potentially recognised by BAG domain
motifSummary(IntAct$res$qslimfinder.Full_IntAct3.FALSE, IntAct_net,
             motif = "L.{0,1}Q.LR", viral_query = NULL, human_seed = NULL)

## Check that human proteins in this subnetwork were used to construst "successful" datasets - yes, they were - human protein ID is copied from "interacts_with" column of QSLIMFinder output which contains a seed protein for the dataset
```

```{r}
# lets compare the number of large-scale studies in human vs viral
# load ppi data
IntAct_ppi = loadIntActFTP(dir = "./data_files/IntActRelease_2017Nov13/",
                                                release = "2017Nov13")
# human-viral
all_viral_interaction_ppi = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                                clean = TRUE, protein_only = TRUE,
                                                MITABdata = IntAct_ppi, directory = "./data_files/",
                                                releaseORdate = "2017Nov13")
# human-human
Full_IntAct_ppi = fullInteractome(taxid = 9606, database = "IntActFTP", format = "tab27",
                              clean = TRUE, protein_only = TRUE,
                              MITABdata = IntAct_ppi, directory = "./data_files/",
                                                releaseORdate = "2017Nov13")


# number of interactions per study
table(all_viral_interaction_ppi$data$Publication_Identifiers)[order(table(all_viral_interaction_ppi$data$Publication_Identifiers), decreasing = T)]
countBaits = function(data){
    viral_ppi_data = copy(data$data)
    viral_prot = viral_ppi_data[, .(prot = c(IDs_interactor_A, IDs_interactor_B), 
                                    bait_prey = c(bait_prey_status_A, bait_prey_status_B),
                                    pmid = c(Publication_Identifiers, Publication_Identifiers))]
    viral_prot = unique(viral_prot)
    viral_prot[, bait_per_study := uniqueN(prot), by = .(pmid, bait_prey)]
    viral_studies = unique(viral_prot[, .(bait_per_study, pmid, bait_prey)])
    viral_studies[order(bait_per_study)][bait_prey == "bait"]
}

viral_studies = countBaits(all_viral_interaction_ppi)
human_studies = countBaits(Full_IntAct_ppi)
```

## Time spend per dataset

```{r}
motifs = fread("/Users/vk7/Desktop/ebi_projects/viral_project/qslimfinder.Full_IntAct3cloudfixF.FALSE/result/main_result.txt", stringsAsFactors = F)
motifs[, time_in_sec := as.integer(as.POSIXct(strptime(RunTime, "%H:%M:%S")))]
motifs[, time_in_sec := time_in_sec - min(time_in_sec)]
dataset_times = unique(motifs[, .(Dataset, time_in_sec, SeqNum)])
hist(dataset_times$time_in_sec)
plot(dataset_times[SeqNum > 170 & SeqNum < 200]$SeqNum, dataset_times[SeqNum > 170 & SeqNum < 200]$time_in_sec)
mean(dataset_times[SeqNum > 170 & SeqNum < 200]$time_in_sec)
```

```{r, exit}
Sys.Date. = Sys.Date()
Sys.Date.
session_info. = devtools::session_info()
session_info.
filename = paste0("./processed_data_files/compr_benchmarking_venn_IntVidVir",format(Sys.Date., "%Y%m"),"_type_", measure1, measure2,".RData")
save(list = ls(), file=filename)
```