---
title: Identify viral-human interactions, full human interactome and download protein
  sequences
author: "Vitalii Kleshchevnikov"
date: "13/07/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages = c("data.table","downloader","R.utils", "PSICQUIC", "ggplot2", "MItools")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
      packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
      install.packages(packages_to_install)
      packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
      source("https://bioconductor.org/biocLite.R")
      biocLite(packages_to_install)
      devtools::install_github("vitkl/viral_project/MItools")
}

suppressPackageStartupMessages({
    library(data.table)
    library(downloader)
    library(R.utils)
    library(PSICQUIC)
    library(ggplot2)
    library(MItools)
})
```

## Seaching for viral-human PPI  

I download all protein-protein interactions from IntAct (IMEx databases) involving viral proteins using PSICQUIC service in MITAB2.7 format (https://psicquic.github.io/MiqlReference27.html). I clean the data in the table to make it more useble. Then, I filter and keep only human-viral interactions.

```{r PSICQUIC_viral_human}
if(!file.exists("./data_files/viral_interactions.txt")){
    queryPSICQUIC(query = "species:10239",
                format = "tab27",
                database = "imex",
                file = "./data_files/viral_interactions.txt")
}

all_viral_interaction = fread("./data_files/viral_interactions.txt", stringsAsFactors = F)

# clean the data in the table to make it more useble
all_viral_interaction = cleanMITAB(all_viral_interaction)

# filter only human-viral interactions
all_viral_interaction = all_viral_interaction[Taxid_interactor_A == "9606" | Taxid_interactor_B == "9606",]
all_viral_interaction = unique(all_viral_interaction)
```

## Searching for human-human PPI

I download all human-human protein-protein interactions from IntAct (IMEx databases) using PSICQUIC service in MITAB2.5 format and clean the data.

```{r PSICQUIC_human_human}
all_human_interaction = fullInteractome(taxid = "9606", database = "imex", format = "tab25",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")

# select proteins
viral_proteins = unique(c(all_viral_interaction[Taxid_interactor_A != "9606", IDs_interactor_A], all_viral_interaction[Taxid_interactor_B != "9606", IDs_interactor_B]))
human_viral_proteins = unique(c(all_viral_interaction[Taxid_interactor_A == "9606", IDs_interactor_A], all_viral_interaction[Taxid_interactor_B == "9606", IDs_interactor_B]))
human_human_proteins = all_human_interaction[,unique(c(IDs_interactor_A,IDs_interactor_B))]
all_proteins = unique(c(viral_proteins, human_viral_proteins, human_human_proteins))
all_proteins = all_proteins[order(all_proteins)]
```

There are `r length(viral_proteins)` viral_proteins (including isoforms and postprocessed chains).  
There are `r length(human_viral_proteins)` human_proteins involved in human-viral interactions (including isoforms and postprocessed chains).  
There are `r length(human_human_proteins)` human_proteins involved in human-human interactions (including isoforms and postprocessed chains).  

## Finding protein sequences

Interacting viral and human proteins belong to three categories:  
- proteins identified by canonical UniProt identifier (P04637): no isoform exist or impossible to distinguish isoforms from the published result  
- proteins identified to an isoform (P04637-1)  
- proteins that are cleaved into the functional fragments from a precursor protein (P04591-PRO_0000261216)  

These categories require different approaches of retrieving sequences.   
1. Retrieving sequences of proteins with canonical UniProt identifier is possible with R package for UniProt webservices or UniProt REST API. These proteins also may not require InterProScan to identify domains.  
2. Isoform sequences are accessible using UniProt REST API: http://www.uniprot.org/uniprot/P04637-2.fasta.
3. Post-processed chains are not straightforward to map as they are defined by the interval of the canonical UniProt sequence (http://www.uniprot.org/uniprot/P04591.fasta). First, we need to retrieve post-processed chain position from Uniprot in gff format: http://www.uniprot.org/uniprot/?query=PRO_0000261216&format=gff. The search return all sequence features from a given protein, we select only post-processed chain we are interested in and then use position specified to subset the sequence.

### Retrieving canonical and isoform sequences

```{r}
# Filtering sequence names by group
canonical_all_proteins = all_proteins[-c(grep("-[[:digit:]]+", all_proteins, value = F),grep("-PRO_[[:digit:]]+$", all_proteins, value = F))]
isoform_all_proteins = grep("-[[:digit:]]+", all_proteins, value = T)

# dowloading FASTA for canonical_and_isoform sequences if file with this data doesn't exist or doesn't contain all sequences
download_fasta(uniprot_ac = canonical_and_isoform, file_name = "./data_files/canonical_and_isoform.fasta")
```

Some protein ids have been UniParc-ed. I will comment the code necessary to download the sequences instead of dealing with the problem as the problem is being generated by de-sync between UniProt and IntAct releases and is, therefore, temporary.  

### Retrieving post-processed chain sequences

```{r}
# Filtering sequence names by group
postproc_all_proteins = grep("-PRO_[[:digit:]]+$", all_proteins, value = T)
postproc = gsub("^[[:alnum:]]+-","",postproc_all_proteins)

# dowloading FASTA for post-processed sequences if file with this data doesn't exist or doesn't contain all sequences
download_fasta_postproc(postproc_id = postproc, file_name = "./data_files/postproc.fasta")
```

### Combining sequences

```{r}
all_human_viral_proteins.fasta = append(readAAStringSet("./data_files/canonical_and_isoform.fasta"), readAAStringSet("./data_files/postproc.fasta"))
writeXStringSet(all_human_viral_proteins.fasta,
                file = "./data_files/all_human_viral_proteins.fasta",
                format="fasta")
```

`r length(all_human_viral_proteins.fasta)` FASTA sequences total.

## Running InterProScan over those sequences

Notes on the installation of the InterProScan: https://github.com/ebi-pf-team/interproscan/wiki/HowToDownload.  
The code below will run InterProScan version 5.25-64.0 over all sequences in all_human_viral_proteins.fasta.   
-i specifies the input file  
-f specifies the output format  
-iprlookup tells to look up InterProID for features (only specific member database ID is returned by default)  
-goterms tells to look up GO terms associated with given InterProID  
-b specifies output file (format suffix like .gff3 will be added automatically)  

InterProScan works on Linux system only. /my_interproscan/interproscan-5.25-64.0/ should be located in dir1 "/dir1/dir2/viral_project"   

```{bash}
# bsub rsync /nfs/research/petsalaki/users/vitalii /hps/nobackup/research/petsalaki/users/ -r -av --delete
../../my_interproscan/interproscan-5.25-64.0/interproscan.sh -i ./data_files/all_human_viral_proteins.fasta -f gff3 -iprlookup -goterms -b ./processed_data_files/all_human_viral_protein_domains
# bsub rsync /hps/nobackup/research/petsalaki/users/vitalii /nfs/research/petsalaki/users/ -r -av
```

```{r}
sessionInfo()
```