---
title: "Motif_search_strategies"
author: "Vitalii Kleshchevnikov"
date: "29/11/2017"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages = c("MItools", "rtracklayer", "R.utils", "ontologyIndex", "parallel")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    # specifying mirror is necessary for some Linux systems
    install.packages(packages_to_install, dependencies = T, repos = "http://mirrors.ebi.ac.uk/CRAN/")
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    source("https://bioconductor.org/biocLite.R")
    biocLite(packages_to_install)
    devtools::install_github("vitkl/MItools", dependencies = T)
}

suppressPackageStartupMessages({
    library(MItools)
    library(rtracklayer)
    library(parallel)
})
```

# Introduction

## Strategies

1. Optimal search settings (3*2):  
- slimlen= 5,8,10 : Maximum length of SLiMs to return (no. non-wildcard positions)  
- maxwild= 2,3 : Maximum number of consecutive wildcard positions to allow  

2. Comparing interaction datasets (4):  
- Full IntAct  
- BioPlex  
- Vidal  
- randomised BioPlex (keeping the degree and the number of interactions intact)  

3. Domain-centered vs protein-centered strategy (2)  

4. QSLIMFinder vs SLIMFinder (2)  

Total: 3 \* 2 \* 4 \* 2 \* 2 = 96 conditions  

This is not feasible, so I will first compare interaction datasets, domain- vs protein- centered strategy and QSLIMFinder vs SLIMFinder (16 conditions).  
I will then test multiple motif settings on a selected dataset (different run of the pipeline, 6 conditions).  

## load PPI data

```{r load_data}
# load ppi data
IntAct = loadIntActFTP(dir = "./data_files/IntActRelease_2017Nov13/")
# human-viral
all_viral_interaction = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                                clean = TRUE, protein_only = TRUE,
                                                MITABdata = IntAct, directory = "./data_files/")
# need to simplify isoform and post-processed chain IDs for human proteins
all_viral_interaction_BioPlex = copy(all_viral_interaction)
all_viral_interaction_BioPlex$data$IDs_interactor_A = gsub("-[[:digit:]]+$","",all_viral_interaction_BioPlex$data$IDs_interactor_A)
all_viral_interaction_BioPlex$data$pair_id = paste0(all_viral_interaction_BioPlex$data$IDs_interactor_A, "|", all_viral_interaction_BioPlex$data$IDs_interactor_B)
all_viral_interaction_BioPlex$data = unique(all_viral_interaction_BioPlex$data)

# human-human
Full_IntAct = fullInteractome(taxid = 9606, database = "IntActFTP", format = "tab27",
                              clean = TRUE, protein_only = TRUE,
                              MITABdata = IntAct, directory = "./data_files/")
BioPlex = loadBioplex(dir = "./data_files/",
                      url = "http://bioplex.hms.harvard.edu/data/BioPlex_2.3_interactionList.tsv")
Vidal = subsetMITABbyPMIDs(MITABdata = Full_IntAct,
                           PMIDs = c("25416956", "unassigned1304"))

randomised_BioPlex = copy(BioPlex)
randomised_BioPlex$data$Publication_Identifiers = NA
randomised_BioPlex$data$Confidence_values = NA
randomised_BioPlex$data = unique(randomised_BioPlex$data)
set.seed(1)
randomised_BioPlex$data$IDs_interactor_B = randomised_BioPlex$data$IDs_interactor_B[sample.int(length(randomised_BioPlex$data$IDs_interactor_B))]
randomised_BioPlex$data$pair_id = paste0(randomised_BioPlex$data$IDs_interactor_A, "|", randomised_BioPlex$data$IDs_interactor_B)
```

## Set standard options

```{r std_options}
# set standard options
myPPInetwork2SLIMFinder = function(dataset_name = "SLIMFinder", slimlen = 5, maxwild = 2,
                                   interaction_main_set, interaction_query_set,
                                   center_domains = F, analysis_type = "qslimfinder"){
    PPInetwork2SLIMFinder(dataset_name = dataset_name,
                          interaction_main_set = interaction_main_set,
                          interaction_query_set = interaction_query_set,
                          analysis_type = analysis_type,
                          options = paste0("dismask=T consmask=F cloudfix=T probcut=0.3 minwild=0 maxwild=",maxwild," slimlen=",slimlen," alphahelix=F maxseq=800 savespace=0 iuchdir=T extras=2"),
                          domain_res_file = "./processed_data_files/what_we_find_VS_ELM_clust20171019.RData",
                          domain_results_obj = "res_count",
                          center_domains = center_domains,
                          fasta_path = "./data_files/all_human_viral_proteins.fasta",
                          main_set_only = F,
                          domain_pvalue_cutoff = 1, # 0.000005 for test run, 1 for full run
                          SLIMFinder_dir = paste0("./",dataset_name,"/"),
                          LSF_project_path = "/hps/nobackup/research/petsalaki/users/vitalii/vitalii/viral_project/",
                          software_path = "../software/cluster/",
                          length_set1_min = 4, # selected based on benchmarking: minimal set size that returns motifs with Sig < 0.05 = 4+1
                          length_set2_min = 1,
                          write_log = T,
                          N_seq = 200,
                          seed_list = NULL, # NULL for full run, seed_list for test run
                          memory_start = 350,
                          memory_step = 2000)
}

# Download ELM data
elm_filename = paste0("/hps/nobackup/research/petsalaki/users/vitalii/vitalii/viral_project/", "./data_files/", Sys.Date(), 
                      "elms_index.tsv")
if (!file.exists(elm_filename)) 
    download.file("http://elm.eu.org/elms/elms_index.tsv", elm_filename)
```

## Define parameters

```{r params}
analysis_type = c("qslimfinder", "slimfinder")
datasets = c("Full_IntAct", "BioPlex", "Vidal", "randomised_BioPlex", "all_viral_interaction")
domain_centered = c(F, T) # c(F, T)
#slimlen = c(5, 8, 10)
#maxwild = c(2, 3)

dataset_names = data.table(analysis_type = character(), datasets = character(),
                           domain_centered = logical(), dataset_names = character(), 
                           slimlen = integer(), maxwild = integer())
for(a in analysis_type){
    for(d in datasets){
        for(cen in domain_centered){
            dataset_names_temp = data.table(analysis_type = a, datasets = d,
                                            domain_centered = cen, dataset_names = paste(a,d,cen, sep = "."), 
                                            slimlen = 5, maxwild = 2)
            dataset_names = rbind(dataset_names, dataset_names_temp)
        }
    }
}
dataset_names[grepl("BioPlex",datasets), query_dataset := "all_viral_interaction_BioPlex"]
dataset_names[!grepl("BioPlex",datasets), query_dataset := "all_viral_interaction"]

# remove already finished names
R_data_names = readLines("./RData_from_Motif_search_strategies")
R_data_names = gsub("\\./processed_data_files/QSLIMFinder_instances_h2v_","", R_data_names)
R_data_names = gsub("_clust201712\\.RData\\.zip","", R_data_names)
R_data_names = gsub("_clust201801\\.RData\\.zip","", R_data_names)
R_data_names = gsub("_clust201802\\.RData\\.zip","", R_data_names)
dataset_names = dataset_names[!dataset_names %in% R_data_names]
```

## Perform motif search

```{r search}
# set up parallel processing
# create cluster
cores = 4
if(n_cores > detectCores() - 1) n_cores = detectCores() - 1
cl <- makeCluster(cores)
# get library support needed to run the code
clusterEvalQ(cl, {library(MItools); library(rtracklayer)})
# put objects in place that might be needed for the code
clusterExport(cl, ls(), envir=environment())

R_data = parSapplyLB(cl = cl, X = 1:nrow(dataset_names), FUN = function(i){
    R_data_name = myPPInetwork2SLIMFinder(dataset_name = dataset_names$dataset_names[i], slimlen = dataset_names$slimlen[i], maxwild = dataset_names$maxwild[i],
                            interaction_main_set = eval(parse(text = dataset_names$datasets[i])),
                            interaction_query_set = eval(parse(text = dataset_names$query_dataset[i])),
                            center_domains = dataset_names$domain_centered[i], analysis_type = dataset_names$analysis_type[i])
    
    zip(zipfile = paste0(R_data_name, ".zip"), files = R_data_name)
    unlink(R_data_name)
    
    R_data_names = readLines("./RData_from_Motif_search_strategies")
    R_data_names = c(R_data_names, paste0(R_data_name, ".zip"))
    write(R_data_names, "./RData_from_Motif_search_strategies")
    
    paste0(R_data_name, ".zip")
}, simplify = TRUE, USE.NAMES = TRUE) # parSapplyLB is parSapply that accounts for different time each dataset might take

stopCluster(cl)

# paths to RData files containing all objects created by all runs of this pipeline
print(R_data)
```