---
title: 'QSLIMFinder: domain instances, human to search, viral to filter'
author: "Vitalii Kleshchevnikov"
date: "04/09/2017"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages = c("MItools", "rtracklayer", "ggplot2", "GGally", "RColorBrewer", "R.utils", "ontologyIndex")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    # specifying mirror is necessary for some Linux systems
    install.packages(packages_to_install, dependencies = T, repos = "http://mirrors.ebi.ac.uk/CRAN/")
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    source("https://bioconductor.org/biocLite.R")
    biocLite(packages_to_install)
    devtools::install_github("vitkl/MItools")
}

suppressPackageStartupMessages({
    library(MItools)
    library(rtracklayer)
    library(ggplot2)
    library(GGally)
    library(RColorBrewer)
    library(scales)
    library(ontologyIndex)
})
```

## use all the interacting partners of human proteins with significant domains to seach for motifs and their viral interactors to filter

QSLIMfinder options: 
- seqin - all the interacting partners of human proteins with significant domains (including viral proteins); FASTA sequences
- query - the viral interacting partners of human proteins with significant domains; names of the FASTA sequences in seqin
- resdir - directory where to store all results
- resfile - file where to write
- dismask - mask disordered regions using IUPRED
- consmask - relative conservation masking
- probcut - FDR-corrected probability cutoff
- qregion - Mask all but the region of the query from (and including) residue X to residue Y [0,-1] (useful for regions sufficient/necessary to interact)

## load PPI data

```{r load_data}
# load ppi data
# human-human
all_human_interaction = fullInteractome(taxid = 9606, database = "IntActFTP", format = "tab27",
                                        clean = TRUE, protein_only = TRUE,
                                        directory = "./data_files/")
# human-viral
all_viral_interaction = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                                clean = TRUE, protein_only = TRUE, 
                                                directory = "./data_files/")
```

## load results ofwhich domains are likely to mediate interaction
```{r load_results}
# load the domain analysis results
load("./processed_data_files/what_we_find_VS_ELM_clust.RData")
domain_res = res
rm(list = ls()[!ls() %in% c("domain_res", "all_human_interaction", "all_viral_interaction")])

# choose pvalue cutoff:
plot(domain_res)
proteins_w_signif_domains = unique(domain_res$data_with_pval[order(p.value, decreasing = F)[1:300], IDs_interactor_human])
```

### prepare sets of protein sequences and instructions for SLIMFinder

```{r prepare_FASTA_for_QSLIMFinder, eval=T}
# load FASTA
all.fasta = readAAStringSet(filepath = "./data_files/all_human_viral_proteins.fasta", format = "fasta")

# remove interactions if one of the interactors does not have a FASTA sequence
all_human_interaction = removeInteractionNoFASTA(all_human_interaction, all.fasta)
all_viral_interaction = removeInteractionNoFASTA(all_viral_interaction, all.fasta)

# all.fasta["Q77M19"] for some reason this name doesn't work for subsetting, that's why I need recodeFASTA()
#all_viral_interaction$data = all_viral_interaction$data[IDs_interactor_A == "O60503",]
#forSLIMFinder = listInteractionSubsetFASTA(interaction_set1 = all_human_interaction, 
#                                           interaction_set2 = all_viral_interaction, 
#                                           seed_id_vect = "O60503",
#                                           fasta = all.fasta,
#                                           single_interact_from_set2 = T, set1_only = T)

forSLIMFinder = listInteractionSubsetFASTA(interaction_set1 = all_human_interaction, 
                                           interaction_set2 = all_viral_interaction, 
                                           seed_id_vect = proteins_w_signif_domains,
                                           fasta = all.fasta,
                                           single_interact_from_set2 = T, set1_only = T)

forSLIMFinder_Ready = filterInteractionSubsetFASTA_list(forSLIMFinder,  length_set1_min = 3, length_set2_min = 1)
# how many viral-human pairs in which viral proteins interact with human protein (with relevant domain) but are not associated with the domain
sum(!domainProteinPairMatch(forSLIMFinder_Ready, domain_res, remove = F))
# how many viral-human pairs in which viral proteins interact with human protein (with relevant domain) and are associated with the domain
sum(domainProteinPairMatch(forSLIMFinder_Ready, domain_res, remove = F))
# total
forSLIMFinder_Ready$length

#filter
domain_filt = domain_res
domain_filt$data_with_pval = domain_filt$data_with_pval[p.value < 0.4,]
forSLIMFinder_Ready = domainProteinPairMatch(forSLIMFinder_Ready, domain_filt, remove = T)

### sanity check
obs = unique(domain_filt$data_with_pval[, .(IDs_interactor_human, IDs_interactor_viral)])
setkey(obs)

res = tstrsplit(gsub("interactors_of\\.|\\.","", names(forSLIMFinder_Ready$fasta_subset_list)), ":")
res = data.table(idA = res[[1]], idB = res[[2]])
res_unkeyed = copy(res)
res = unique(res[complete.cases(res),])
setkey(res)

mean(paste0(res$idA, res$idB) %in% paste0(obs$IDs_interactor_human, obs$IDs_interactor_viral))

query_in_FASTA = sapply(1:length(forSLIMFinder_Ready$fasta_subset_list), function(i, forSLIMFinder){
    res = tstrsplit(gsub("interactors_of\\.|\\.","", names(forSLIMFinder$fasta_subset_list)), ":")
    res[[2]][i] %in% names(forSLIMFinder$fasta_subset_list[[i]])
}, forSLIMFinder_Ready)
mean(query_in_FASTA)
### sanity check end
```

## Generate bash commands to run QSLIMFinder

```{r save}
SLIMFinder_dir = "./SLIMFinder/"
if(!dir.exists(SLIMFinder_dir)) dir.create(SLIMFinder_dir)

forSLIMFinder_file_list = writeInteractionSubsetFASTA_list(interactionFASTA_list = forSLIMFinder_Ready,
                                                           dir = "./SLIMFinder/")

QSLIMFinderCommand(file_list = forSLIMFinder_file_list)

all_commands = mQSLIMFinderCommand(file_list = forSLIMFinder_file_list, 
                                   slimpath = "../software/cluster/slimsuite/tools/qslimfinder.py", 
                                   blast = "../software/cluster/ncbi_blast_2.6.0/bin/", 
                                   iupred = "../software/cluster/iupred/iupred",
                                   options = "dismask=T consmask=F cloudfix=T probcut=0.3 iuchdir=T",
                                   LSF_cluster_mode = T,
                                   LSF_project_path = "/hps/nobackup/research/petsalaki/users/vitalii/vitalii/viral_project/",
                                   LSF_cluster_par = "bsub -n 1 -q research-rh7 -M 100 -R \"rusage[mem=100]\"")
```

## Run QSLIMFinder

```{r run}
if(tryCatch({system("bjobs", intern = T)}, error = function(e) e)$message != "error in running command") onLSF = T else onLSF = F
runQSLIMFinder(commands = all_commands, onLSF = onLSF)
```

## Read results

```{r read}
readQSLIMFinderMain = function(file_list = "forSLIMFinder_file_list"){
    main_result = lapply(file_list$outputfile, function(file) {
        #fread("./SLIMFinder/output/interactors_of.O60506_P03496./main_result", stringsAsFactors = F)
        if(file.exists(file)) fread(file, stringsAsFactors = F)
    })
    Reduce(rbind, main_result)
}
writePatternList = function(QSLIMFinder_main_result, filename = "./motifs.txt") {
    motifs = unique(QSLIMFinder_main_result[,.(Dataset,Pattern)])
    motifs[, Name := paste0(Dataset,Pattern)][,Dataset:=NULL]
    fwrite(motifs, filename, sep = "\t")
}

QSLIMFinder_main_result = readQSLIMFinderMain(file_list = forSLIMFinder_file_list)
writePatternList(QSLIMFinder_main_result, filename = "./SLIMFinder/result/motifs.txt")
```

## Compare motifs to ELM

```{r comparimotif}
runCompariMotif3 = function(input_file = "./SLIMFinder/result/motifs.txt",
                            slimpath = "../software/slimsuite/tools/",
                            dbpath = "./data_files/",
                            dburl = "http://elm.eu.org/elms/elms_index.tsv",
                            parameters = "unmatched=T motific=T", 
                            out_file = "./SLIMFinder/result/comparimotif.tdt",
                            LSF_project_path = "/hps/nobackup/research/petsalaki/users/vitalii/vitalii/viral_project/",
                            run = F){
    elm_filename = paste0(dbpath, Sys.Date(), "elms_index.tsv")
    if(!file.exists(elm_filename)) download.file(dburl, elm_filename)
    comparimotif_call = paste0("python ", LSF_project_path, slimpath, "comparimotif_V3.py")
    input_res = paste0("motifs=", LSF_project_path, input_file)
    input_database = paste0("searchdb=", LSF_project_path, elm_filename) 
    out_file = paste0("resfile=", LSF_project_path, out_file)
    command = paste(comparimotif_call, input_res, input_database, parameters, out_file)
    if(run) system(command, wait = F)
    return(command)
}
runCompariMotif3(slimpath = "../software/cluster/slimsuite/tools/",
                 run = T)
```