---
title: 'QSLIMFinder: domain instances, human to search, viral to filter'
author: "Vitalii Kleshchevnikov"
date: "04/09/2017"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages = c("MItools", "rtracklayer", "ggplot2", "GGally", "RColorBrewer", "R.utils")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    # specifying mirror is necessary for some Linux systems
    install.packages(packages_to_install, dependencies = T, repos = "http://mirrors.ebi.ac.uk/CRAN/")
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    source("https://bioconductor.org/biocLite.R")
    biocLite(packages_to_install)
    devtools::install_github("vitkl/MItools")
}

suppressPackageStartupMessages({
    library(MItools)
    library(rtracklayer)
    library(ggplot2)
    library(GGally)
    library(RColorBrewer)
})
```

## use all the interacting partners of human proteins with significant domains to seach for motifs and their viral interactors to filter

QSLIMfinder: 
- seqin - all the interacting partners of human proteins with significant domains (including viral proteins); FASTA sequences
- query - the viral interacting partners of human proteins with significant domains; names of the FASTA sequences in seqin
- resdir - directory where to store all results
- resfile - file where to write
- dismask - mask disordered regions using IUPRED
- consmask - relative conservation masking
- probcut - FDR-corrected probability cutoff
- qregion - Mask all but the region of the query from (and including) residue X to residue Y [0,-1] (useful for regions sufficient/necessary to interact)



### prepare sets of protein sequences and instructions for SLIMFinder

```{r}
# load the domain analysis results
load("./processed_data_files/what_we_find_VS_ELM_output_freq.RData")

# load ppi data
# human-viral
all_viral_interaction = queryPSICQUICrlib(query = "species:10239",
                format = "tab27",
                database = "imex",
                directory = "./data_files/")
all_viral_interaction = cleanMITAB(all_viral_interaction)
# pick proteins only
all_viral_interaction$data = all_viral_interaction$data[interactor_IDs_databases_A == "uniprotkb" & interactor_IDs_databases_B == "uniprotkb", ]
# filter only human-viral interactions
all_viral_interaction$data = all_viral_interaction$data[Taxid_interactor_A == 9606 | Taxid_interactor_B == 9606,]
all_viral_interaction$data = unique(all_viral_interaction$data)
# human-human
all_human_interaction = fullInteractome(taxid = 9606, database = "IntActFTP", format = "tab27",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")
# human-viral
all_viral_interactionInt = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")
all_viral_interactionImex = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "imex", format = "tab27",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")


full_interactome$data[grepl("D1LN35", `#ID(s) interactor A`) & grepl("Q9HAV7", `ID(s) interactor B`),]
full_interactome_clean$data[IDs_interactor_A == "D1LN35" & IDs_interactor_B == "Q9HAV7",]

mitab[grepl("D1LN35", IDs_interactor_A) & grepl("Q9HAV7", IDs_interactor_B),]
mitab[IDs_interactor_A == "D1LN35" & IDs_interactor_B == "Q9HAV7",]


all_viral_interactionInt_RAW = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                        clean = F, protein_only = F, directory = "./data_files/")
all_viral_interactionImex_RAW = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "imex", format = "tab27",
                                        clean = F, protein_only = F, directory = "./data_files/")

all_viral_interactionImex$data[!(all_viral_interactionImex$data[,paste0(.SD, collapse = ""),
                                                                by = 1:nrow(all_viral_interactionImex$data),
                                                                .SDcols = colnames(all_viral_interactionInt$data)[c(1:8, 20:26)]]$V1 %in%
                                     all_viral_interactionInt$data[,paste0(.SD, collapse = ""),
                                                                   by = 1:nrow(all_viral_interactionInt$data),
                               .SDcols = colnames(all_viral_interactionInt$data)[c(1:8, 20:26)]]$V1)]

all_viral_interactionInt$data[!(all_viral_interactionInt$data[,paste0(.SD, collapse = ""),
                                                                by = 1:nrow(all_viral_interactionInt$data),
                                                                .SDcols = colnames(all_viral_interactionInt$data)[c(1:8, 20:26)]]$V1 %in%
                                     all_viral_interactionImex$data[,paste0(.SD, collapse = ""),
                                                                   by = 1:nrow(all_viral_interactionImex$data),
                               .SDcols = colnames(all_viral_interactionImex$data)[c(1:8, 20:26)]]$V1)]

colnames(all_viral_interactionImex$data)
colnames(all_viral_interactionInt$data)
# load FASTA
all.fasta = readAAStringSet(filepath = "./data_files/all_human_viral_proteins.fasta", format = "fasta")

SLIMFinder_dir = "./SLIMFinder/"
if(!dir.exists(SLIMFinder_dir)) dir.create(SLIMFinder_dir)
fwrite(fread("./SLIMFinder/filelist.txt"), "./SLIMFinder/filelist.txt", sep = "\t")

E3ligases_url = "http://www.uniprot.org/uniprot/?query=name%3Ae3+name%3Aligase+organism:9606&format=tab&columns=id"
E3ligasesinteractions = subsetMITABbyID(all_human_interaction, ID_seed = fread(E3ligases_url, stringsAsFactors = F)$Entry, within_seed = F, only_seed2nonseed = T)

```

## Run QSLIMFinder 

```{bash}
software_PATH="../software/"
SLIM_PATH=$software_PATH"slimsuite/tools/"
QSLIMFinder=$SLIM_PATH"qslimfinder.py"
IUPred_PATH=$software_PATH"iupred/iupred"
BLAST_PATH=$software_PATH"ncbi_blast_2.6.0/bin/"

res_dir="./SLIMFinder/QSLIMFinder/"
res_path=$res_dir"res.csv"
echo $QSLIMFinder
echo $IUPred_PATH
#python $QSLIMFinder 

#QSLIMFinder
while IFS=$'\t' read -r data query;
do
 python $QSLIMFinder blast+path=$BLAST_PATH iupath=$IUPred_PATH resdir=$res_dir resfile=$res_path seqin=$data query=$query dismask=T consmask=T cloudfix=F probcut=1
done < ./SLIMFinder/filelist.txt

res_dir="./SLIMFinder/QSLIMFinder/"
res_path=$res_dir"SLIMFinder_res.csv"
SLIMFinder=$SLIM_PATH"slimfinder.py"
#SLIMFinder
while IFS=$'\t' read -r data query;
do
 python $SLIMFinder blast+path=$BLAST_PATH iupath=$IUPred_PATH resdir=$res_dir resfile=$res_path seqin=$data dismask=T cloudfix=F probcut=1
done < ./SLIMFinder/filelist.txt
```


