---
title: 'QSLIMFinder: domain instances, human to search, viral to filter'
author: "Vitalii Kleshchevnikov"
date: "04/09/2017"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages = c("MItools", "rtracklayer", "ggplot2", "GGally", "RColorBrewer", "R.utils")
if(mean(packages %in% names(installed.packages()[,"Package"])) != 1){
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    # specifying mirror is necessary for some Linux systems
    install.packages(packages_to_install, dependencies = T, repos = "http://mirrors.ebi.ac.uk/CRAN/")
    packages_to_install = packages[!packages %in% names(installed.packages()[,"Package"])]
    source("https://bioconductor.org/biocLite.R")
    biocLite(packages_to_install)
    devtools::install_github("vitkl/MItools")
}

suppressPackageStartupMessages({
    library(MItools)
    library(rtracklayer)
    library(ggplot2)
    library(GGally)
    library(RColorBrewer)
})
```

## use all the interacting partners of human proteins with significant domains to seach for motifs and their viral interactors to filter

QSLIMfinder: 
- seqin - all the interacting partners of human proteins with significant domains (including viral proteins); FASTA sequences
- query - the viral interacting partners of human proteins with significant domains; names of the FASTA sequences in seqin
- resdir - directory where to store all results
- resfile - file where to write
- dismask - mask disordered regions using IUPRED
- consmask - relative conservation masking
- probcut - FDR-corrected probability cutoff
- qregion - Mask all but the region of the query from (and including) residue X to residue Y [0,-1] (useful for regions sufficient/necessary to interact)



### prepare sets of protein sequences and instructions for SLIMFinder

```{r}
# load the domain analysis results
load("./processed_data_files/what_we_find_VS_ELM_output_freq.RData")
domain_res = res

# load ppi data
# human-human
all_human_interaction = fullInteractome(taxid = 9606, database = "IntActFTP", format = "tab27",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")
# human-viral
all_viral_interaction = interSpeciesInteractome(taxid1 = 9606, taxid2 = 10239, database = "IntActFTP", format = "tab27",
                                        clean = TRUE, protein_only = TRUE, directory = "./data_files/")
# load FASTA
all.fasta = readAAStringSet(filepath = "./data_files/all_human_viral_proteins.fasta", format = "fasta")

subsetMITABbyID(all_human_interaction, ID_seed,
                within_seed = F, only_seed2nonseed = T)
subsetMITABbyID(all_viral_interaction, ID_seed,
                within_seed = F, only_seed2nonseed = T)
AAStringSetList(all = all.fasta)[[1]]

SLIMFinder_dir = "./SLIMFinder/"
if(!dir.exists(SLIMFinder_dir)) dir.create(SLIMFinder_dir)
fwrite(fread("./SLIMFinder/filelist.txt"), "./SLIMFinder/filelist.txt", sep = "\t")



```

## Run QSLIMFinder 

```{bash}
software_PATH="../software/"
SLIM_PATH=$software_PATH"slimsuite/tools/"
QSLIMFinder=$SLIM_PATH"qslimfinder.py"
IUPred_PATH=$software_PATH"iupred/iupred"
BLAST_PATH=$software_PATH"ncbi_blast_2.6.0/bin/"

res_dir="./SLIMFinder/QSLIMFinder/"
res_path=$res_dir"res.csv"
echo $QSLIMFinder
echo $IUPred_PATH
#python $QSLIMFinder 

#QSLIMFinder
while IFS=$'\t' read -r data query;
do
 python $QSLIMFinder blast+path=$BLAST_PATH iupath=$IUPred_PATH resdir=$res_dir resfile=$res_path seqin=$data query=$query dismask=T consmask=T cloudfix=F probcut=1
done < ./SLIMFinder/filelist.txt

res_dir="./SLIMFinder/QSLIMFinder/"
res_path=$res_dir"SLIMFinder_res.csv"
SLIMFinder=$SLIM_PATH"slimfinder.py"
#SLIMFinder
while IFS=$'\t' read -r data query;
do
 python $SLIMFinder blast+path=$BLAST_PATH iupath=$IUPred_PATH resdir=$res_dir resfile=$res_path seqin=$data dismask=T cloudfix=F probcut=1
done < ./SLIMFinder/filelist.txt
```


